example/broker.cpp

```diff
@@ -66,14 +66,41 @@ class server_no_tls {
 
 #if defined(MQTT_USE_TLS)
 
+bool verify_certificate(std::string const &verify_field, bool preverified, boost::asio::ssl::verify_context& ctx, std::shared_ptr<MQTT_NS::optional<std::string>> const& username) {
+    if (!preverified) return false;
+    int error = X509_STORE_CTX_get_error(ctx.native_handle());
+    if (error != X509_V_OK) {
+        int depth = X509_STORE_CTX_get_error_depth(ctx.native_handle());
+        MQTT_LOG("mqtt_broker", error) << "Certificate validation failed, depth: " << depth << ", message: " << X509_verify_cert_error_string(error);
+        return false;
+    }
+
+    X509* cert = X509_STORE_CTX_get_current_cert(ctx.native_handle());
+    X509_NAME* name = X509_get_subject_name(cert);
+
+    std::string verify_field_value;
+    auto obj = std::unique_ptr<ASN1_OBJECT, decltype(&ASN1_OBJECT_free)>(OBJ_txt2obj(verify_field.c_str(), 0), &ASN1_OBJECT_free);
+    if (obj) { // return nullptr if error
+        verify_field_value.resize(MQTT_NS::broker::max_cname_size);
+        auto size = X509_NAME_get_text_by_OBJ(name, obj.get(), &verify_field_value[0], static_cast<int>(verify_field_value.size()));
+        // Size equals -1 if field is not found, otherwise, length of value
+        verify_field_value.resize(static_cast<std::size_t>(std::max(size, 0)));
+    }
+
+    MQTT_LOG("mqtt_broker", info) << "[clicrt] " << verify_field << ":" << verify_field_value;
+    *username = verify_field_value;
```

Comment:
```suggestion
    auto obj = std::unique_ptr<ASN1_OBJECT, decltype(&ASN1_OBJECT_free)>(OBJ_txt2obj(verify_field.c_str(), 0), &ASN1_OBJECT_free);
    if (obj) { // return nullptr if error
        verify_field_value.resize(MQTT_NS::broker::max_cname_size);
        auto size = X509_NAME_get_text_by_OBJ(name, obj.get(), &verify_field_value[0], static_cast<int>(verify_field_value.size()));
        // Size equals -1 if field is not found, otherwise, length of value
        verify_field_value.resize(static_cast<std::size_t>(std::max(size, 0)));
        MQTT_LOG("mqtt_broker", info) << "[clicrt] " << verify_field << ":" << verify_field_value;
        *username = verify_field_value;
    }
```
I'm not 100% sure my suggestion is good.

If non exist field is set, then obj is NULL.
ASN1_OBJECT_free(NULL) is called but it is no problem.
https://www.openssl.org/docs/man1.1.1/man3/ASN1_OBJECT_free.html
> ASN1_OBJECT_free() frees up the ASN1_OBJECT structure a. If a is NULL, nothing is done.

However, even if obj is NULL, `*username = verify_fied_value` is executed.
So *username is updated from nullopt to empty string.
I think that keep it nullopt is appropriate. 
What do you think?

Reply:
I think suggestion is good